src/with_const_generics.rs

```diff
@@ -241,16 +241,283 @@ impl<'a, T, const CAP: usize> IntoIterator for &'a mut ConstGenericRingBuffer<T,
+        // assertion: we can never be in a situation where we have to write more than a batch size of items
+        debug_assert!(
+            p1.len() + p2.len() <= BATCH_SIZE,
+            "p1: {}; p2: {}; batch: {}",
+            p1.len(),
+            p2.len(),
+            BATCH_SIZE
+        );
+
+        // if we are lucky, we're not on the boundary so either p1 or p2 has a length of Self::BATCH_SIZE
+        if p1.len() == BATCH_SIZE {
+            for (index, i) in data.into_iter().enumerate() {
+                p1[index] = MaybeUninit::new(i);
+            }
+        } else if p2.len() == BATCH_SIZE {
+            for (index, i) in data.into_iter().enumerate() {
+                p2[index] = MaybeUninit::new(i);
+            }
+        } else {
+            // oof, unfortunately we're on a boundary
+
+            // iterate over the data
+            let mut data_iter = data.into_iter();
+
+            // put p1.len() in p1
+            for i in p1 {
+                let next_item = data_iter.next();
+                // Safety: p1.len() + p2.len() <= Self::BATCH_SIZE so the two loops here
+                // together cannot run for more than Self::BATCH_SIZE iterations
+                *i = MaybeUninit::new(unsafe { next_item.unwrap_unchecked() });
+            }
+
+            // put p2.len() in p2
+            for i in p2 {
+                let next_item = data_iter.next();
+                // Safety: p1.len() + p2.len() <= Self::BATCH_SIZE so the two loops here
+                // together cannot run for more than Self::BATCH_SIZE iterations
+                *i = MaybeUninit::new(unsafe { next_item.unwrap_unchecked() });
+            }
+        }
+    }
+
+    #[inline]
+    #[cfg(feature = "batched_extend")]
+    fn fill_batch<const BATCH_SIZE: usize>(
+        batch: &mut [MaybeUninit<T>; BATCH_SIZE],
+        iter: &mut impl Iterator<Item = T>,
+    ) -> usize {
+        for (index, b) in batch.iter_mut().enumerate() {
+            if let Some(i) = iter.next() {
+                *b = MaybeUninit::new(i);
+            } else {
+                return index;
+            }
+        }
+
+        BATCH_SIZE
+    }
+
+    #[inline]
+    #[cfg(feature = "batched_extend")]
+    fn extend_batched_internal<const BATCH_SIZE: usize>(
+        &mut self,
+        mut other: impl Iterator<Item = T>,
+    ) {
+        // SAFETY: if CAP < Self::BATCH_SIZE we can't run extend_from_arr_batch so we catch that here
+        if CAP < BATCH_SIZE {
+            for i in other {
+                self.push(i);
+            }
+        } else {
+            // Safety: assume init to MaybeUninit slice is safe
+            let mut batch: [MaybeUninit<T>; BATCH_SIZE] =
+                unsafe { MaybeUninit::uninit().assume_init() };
+
+            // repeat until we find an empty batch
+            loop {
+                // fill up a batch
+                let how_full = Self::fill_batch(&mut batch, &mut other);
```

Comment:
I think the code might be simpler if you fill the batch with `min(BATCH_SIZE, CAP - len)` instead. Then you never have to split the batch up and stuff.

If I understand correctly, `finish_iter` is necessary because you need the writeptr not to be equal to readptr or something? _Maybe_ if you do the `min` thing, that might not be necessary? In any case, might be nice to document why and in what cases you need `finish_iter`.

Reply:
No wait, I mean `min(BATCH_SIZE, CAP - mod(writeptr, CAP))` I think. Whatever the number is of elements that fit until the end of the allocated space. So you don't have to split.